gcc/rust/checks/errors/rust-unsafe-checker.cc

```diff
@@ -784,7 +784,25 @@ UnsafeChecker::visit (Trait &trait)
 void
 UnsafeChecker::visit (ImplBlock &impl)
 {
-  // FIXME: Handle unsafe impls
+  bool safe = !impl.is_unsafe ();
+  // Check for unsafe-only attributes on generics and lifetimes
+  if (safe)
+    for (auto &parm : impl.get_generic_params ())
+      {
+	if (!parm->has_outer_attribute ())
+	  continue;
+
+	AST::Attribute &attr = parm->get_outer_attribute ();
+
+	AST::SimplePath path = attr.get_path ();
+
+	if (path != Values::Attributes::MAY_DANGLE)
+	  continue;
+
+	rust_error_at (attr.get_locus (), ErrorCode::E0133,
+		       "use of %<may_dangle%> unsafe and requires unsafe impl");
```

Comment:
I would switch the logic around to better show the intent, but this is completely valid :)

```suggestion
	if (path == Values::Attributes::MAY_DANGLE)
	rust_error_at (attr.get_locus (), ErrorCode::E0133,
		       "use of %<may_dangle%> is unsafe and requires unsafe impl");
```